### PR TITLE
HIVE-15540: Impl `DatabaseMetaData#getURL()` and `DatabaseMetaData#getUserName()` for HiveServer2 JDBC Driver

### PR DESCRIPTION
**Pull request by linghengqian**
### What changes were proposed in this pull request?

- Impl `DatabaseMetaData#getURL()` and `DatabaseMetaData#getUserName()` for HiveServer2 JDBC Driver.

### Why are the changes needed?

- I totally don't understand why `DatabaseMetaData#getURL()` and `DatabaseMetaData#getUserName()` are not implemented for HiveServer2 JDBC Driver in https://github.com/apache/shardingsphere/pull/33762 and https://github.com/apache/shardingsphere/issues/29052 . This leads to funny logic like the following.

```java
        try (Connection connection = dataSource.getConnection()) {
            return connection.getMetaData().getURL();
        } catch (final SQLFeatureNotSupportedException sqlFeatureNotSupportedException) {
            try (Connection connection = dataSource.getConnection()) {
                Class<?> hiveConnectionClass = Class.forName("org.apache.hive.jdbc.HiveConnection");
                if (connection.isWrapperFor(hiveConnectionClass)) {
                    Object hiveConnection = connection.unwrap(hiveConnectionClass);
                    String connectedUrl = (String) hiveConnectionClass.getMethod("getConnectedUrl").invoke(hiveConnection);
                    assertThat(connectedUrl, is("jdbc:hive2://127.0.0.1:2181/demo_ds_0;initFile=/tmp/init.sql;serviceDiscoveryMode=zooKeeper;zooKeeperNamespace=hiveserver2"));
                    return connectedUrl;
                }
                throw new SQLWrapperException(sqlFeatureNotSupportedException);
            } catch (final SQLException | ClassNotFoundException | NoSuchMethodException | InvocationTargetException | IllegalAccessException exception) {
                throw new RuntimeException(exception);
            }
        } catch (final SQLException ex) {
            throw new SQLWrapperException(ex);
        }
```

### Does this PR introduce _any_ user-facing change?

- Maybe only third-party frameworks or tools like DBeaver will find the impact of the current PR. If one uses a web framework like Spring Boot framework, he has no way of knowing what the current PR does.

### Is the change a dependency upgrade?

- No

### How was this patch tested?

```shell
sdk install java 8.0.422-tem
sdk use java 8.0.422-tem
sdk install maven
mvn clean install -DskipTests -Pitests
mvn test -Dtest=TestHiveDatabaseMetaData -Dtest.output.overwrite=true -pl itests/hive-unit -Pitests
```

### CODE REVIEW

**Review by InvisibleProgrammer — CHANGES_REQUESTED** [review]

File: jdbc/src/java/org/apache/hive/jdbc/HiveDatabaseMetaData.java [hunk 1]
```diff
@@ -744,12 +744,12 @@ public <T> T getObject(int columnIndex, Class<T> type) throws SQLException {
...
 
-  public String getURL() throws SQLException {
-    throw new SQLFeatureNotSupportedException("Method not supported");
+  public String getURL() {
+    return connection.getConnectedUrl();
   }
 
   public String getUserName() throws SQLException {
-    throw new SQLFeatureNotSupportedException("Method not supported");
+    return connection.getConnParams().getSessionVars().get("user");
```
[the remark this answers is not in the document]
Comment (linghengqian) [reply jdbc/src/java/org/apache/hive/jdbc/HiveDatabaseMetaData.java:752]: - `HiveConnection#getSessionValue(String, String)` and `HiveConnection#getUserName()` are both private methods. I guess what you mean is?
```java
  public String getUserName() throws SQLException {
    String username = connection.getConnParams().getSessionVars().get(JdbcConnectionParams.AUTH_USER);
    if ((username == null) || username.isEmpty()) {
      username = JdbcConnectionParams.ANONYMOUS_USER;
    }
    return username;
  }
```
- Changing the `org.apache.hive.jdbc.TestJdbcDriver` doesn't seem intuitive enough, should I create a `org.apache.hive.jdbc.TestHiveDatabaseMetaData` class?
- By the way, Jenkins used by Hive limits a single PR to 5 hours between two CIs, and the current PR's github tag is a bit inconsistent with the actual situation.

[the remark this answers is not in the document]
Comment (InvisibleProgrammer) [reply jdbc/src/java/org/apache/hive/jdbc/HiveDatabaseMetaData.java:752]: > HiveConnection#getSessionValue(String, String) and HiveConnection#getUserName() are both private methods. I guess what you mean is?

Yes, something like that. 

> Changing the org.apache.hive.jdbc.TestJdbcDriver doesn't seem intuitive enough, should I create a org.apache.hive.jdbc.TestHiveDatabaseMetaData class?
Excuse me, wrong copy-paste. So, there is already a test class here: https://github.com/apache/hive/blob/master/itests/hive-unit/src/test/java/org/apache/hive/jdbc/TestHiveDatabaseMetaData.java

[the remark this answers is not in the document]
Comment (linghengqian) [reply jdbc/src/java/org/apache/hive/jdbc/HiveDatabaseMetaData.java]: - I added new unit tests. I guess Hive Doc does not prohibit the use of Guava's utility classes in unit tests.
- I must admit that I found the Hive Contributor Guide on ASF Confluence to be out of date, and I had to spend some time looking up how to start a single unit test locally.
```shell
sdk install java 8.0.422-tem
sdk use java 8.0.422-tem
sdk install maven
mvn clean install -DskipTests -Pitests
mvn test -Dtest=TestHiveDatabaseMetaData -Dtest.output.overwrite=true -pl itests/hive-unit -Pitests
```

**Review by InvisibleProgrammer — APPROVED** [review]
LGTM. Thank you for the change.

**Review by linghengqian — COMMENTED** [review]
- I just did a quick check of https://ci.hive.apache.org/blue/organizations/jenkins/hive-precommit/activity and found that unit tests for `TestMiniLlapLocalCliDriver` are failing in all recent PRs. Is there a quicker way to complete the tests mentioned in https://ci.hive.apache.org/blue/organizations/jenkins/hive-precommit/detail/PR-5554/2/tests than the following command?
```shell
sdk install java 8.0.422-tem
sdk use java 8.0.422-tem
sdk install maven
mvn clean install -DskipTests -Pitests
mvn test -Dtest=TestMiniLlapLocalCliDriver -Dtest.output.overwrite=true -pl itests/qtest -Pitests
```
- Update in 2024.11.24: See https://github.com/apache/hive/pull/5555 .

File: itests/hive-unit/src/test/java/org/apache/hive/jdbc/TestHiveDatabaseMetaData.java [hunk 2]
```diff
@@ -18,18 +18,15 @@
...
+import java.sql.SQLException;
+import java.util.HashMap;
 import java.util.LinkedHashMap;
-import java.util.Properties;
 import java.util.Map;
-import java.util.HashMap;
-import java.sql.SQLException;
-
-import org.junit.Before;
-import org.junit.Test;
```
Comment (ayushtkn) [inline itests/hive-unit/src/test/java/org/apache/hive/jdbc/TestHiveDatabaseMetaData.java:32]: Can you just avoid changing the import order, just makes things complicated while backporting

File: jdbc/src/java/org/apache/hive/jdbc/HiveDatabaseMetaData.java [hunk 3]
```diff
@@ -744,12 +744,16 @@ public <T> T getObject(int columnIndex, Class<T> type) throws SQLException {
...
   }
 
   public String getUserName() throws SQLException {
-    throw new SQLFeatureNotSupportedException("Method not supported");
+    String userName = connection.getConnParams().getSessionVars().get(JdbcConnectionParams.AUTH_USER);
+    if ((userName == null) || userName.isEmpty()) {
+      userName = JdbcConnectionParams.ANONYMOUS_USER;
+    }
+    return userName;
   }
```
Comment (ayushtkn) [inline jdbc/src/java/org/apache/hive/jdbc/HiveDatabaseMetaData.java:757]: can we just do
```
  public String getUserName() throws SQLException {
    return connection.getUserName();
  }
```

and make ``getUserName()`` in ``HiveConnection`` package-private (default)?

**Review by ayushtkn — COMMENTED** [review]
Minor comments, rest looks good

File: jdbc/src/java/org/apache/hive/jdbc/HiveDatabaseMetaData.java (hunk 3 above)
Comment (linghengqian) [reply jdbc/src/java/org/apache/hive/jdbc/HiveDatabaseMetaData.java]: Done.

File: itests/hive-unit/src/test/java/org/apache/hive/jdbc/TestHiveDatabaseMetaData.java (hunk 2 above)
Comment (linghengqian) [reply itests/hive-unit/src/test/java/org/apache/hive/jdbc/TestHiveDatabaseMetaData.java]: Done.

**Review by ayushtkn — APPROVED** [review]
LGTM

**Pull request merged**